### PR TITLE
Pr sagetech mxs

### PR DESCRIPTION
**Pull request by faberc**
# Sagetech MXS PX4 Support Driver

This pull request provides support for Sagetech MXS transponders.

__Authors:__ @mmccormick12 and @faberc.

## Testing

This code has been tested in SITL, on a Cubepilot CubeOrange flight controller, and a CUAV X7+ flight controller.

https://user-images.githubusercontent.com/31755523/173162764-6d236fa0-4901-4c68-a90f-4efaba893c8d.mp4

## Thanks!

Thank you very much to Ryan Johnston and @dakejahl as well as others in the PX4 Slack for your assistance. Feedback on this pull request would be very appreciated.

## Getting And Building this Code

Clone the Sagetech Avionics fork source code and switch to the branch for this pull request.

```
git clone --recursive https://github.com/Sagetech-Avionics/PX4-Autopilot.git
cd PX4_Autopilot
git switch pr-sagetech-mxs
```

To build the code follow the typical PX4 build instructions for your flight controller. This driver must be enabled in the px4board file for your board.

## Issues

- [ ] Flight IDs for vehicles begin to disappear after a while
- [ ] Sometimes causes repeated restarting of flight controller before settling down and fully connecting
- [ ] A few instances observed wehre no vehicles except ownship (without FlightId and altitude) showing up until reboot
- [ ] Want to implement a vehicle timeout/replacement scheme to replace inactive vehicles with newer vehicles
- [ ] Add transponder and system failure message handling.
- [ ] Add ackID mismatch handling.
- [ ] Remove commented out code

---

# Using Sagetech MXS Transponder with PX4

SagetechLogo

## Required Hardware

- Sagetech MXS Certified Transponder
- Flight Controller Board (i.e. CubeOrange or CUAV X7+)
- GPS

By default, this guide is assuming that you are connecting the MXS transponder to your flight controller through the MXS COM 1 port via a serial to TTL adapter.

## Example Connection

mxs_px4_connection

## Step 1: Configure Serial Parameters

Identify which serial port you will be connecting the MXS to.

- __MXS_SER_PORT__: Select the port on your board that the MXS is connected to. (example: TELEM 2).
- __SER_MXS_BAUD__: Select the baud rate you plan on communicating to the MXS with. By default the baud is 57600.

Reboot of the flight controller is required in order for changes to these settings to take effect.

## Step 2: Configure the MXS

The following parameters allow configuration of the MXS before use.

- __MXS_TARG_PORT__: Select the port on the MXS that is connected to your flight controller (example: COM1).
- __MXS_EXT_CFG__: MXS external configuration. If enabled, MXS must be configured externally through Sagetech Mission Control Software. (see [Manual Configuration](#manual-configuration) section) By default this is disabled, and conducts automatic configuration of the MXS.

Reboot of the flight controller is required in order for changes to these settings to take effect.

## Step 3: Configure ADSB Parameters

The following parameters allow configuration of the ADSB-out information sent from your vehicle identifying it and its capabilities to other vehicles with ADSB capability.

- __ADSB_EMERGC__: Set the emergency type. Default: NoEmergency.
- __ADSB_ICAO_ID__: Set the unique vehicle identification number of your aircraft. If set to 0, a random value will be generated. Transponder must be in OFF mode to set the ICAO.
- __ADSB_EMIT_TYPE__: ADSB classification of vehicle type. Default values is 14 (UAV)
- __ADSB_LEN_WIDTH__: Aircraft length and width dimension options in Length and Width in meters. In most cases, use a value of 1 for smallest size.
- __ADSB_MAX_SPEED__: The maximum speed in knots of the vehicle.
- __ADSB_SQUAWK__: Squawk/Transponder (Mode 3/A) code that is braodcasted to ATC that is usually assigned by your ATC for a given flight. In the USA/Canada the default squawk code is for VFR which is 1200. Most parts of Europe and Australia use 7000. If an invalid octal number is set then it will be reset to 1200.
- __ADSB_LIST_MAX__: The number of vehicles to keep track of. Default: 25.
- __ADSB_ICAO_SPECL__ : Special ICAO ID to attempt to track at all times.

Reboot of the flight controller is required in order for changes to these settings to take effect.

## Usage

If __MXS_EXT_CFG__ is disabled, if all of the hardware is correctly connected, after step 3, the user can simply open up QGroundControl and see ADSB traffic if there is a valid GPS fix.

### Commands

The following commands may be sent to the driver during runtime in order to change certain configurations using the following commands:
 - `sagetech_mxs flightid <flightid>` : The user may change the flight ID in flight by providing an 8-character `<flightid>` using this command.
 - `sagetech_mxs opmode <opmode>` : The user can control the transponder operating mode. Valid opmodes include: ['off', 'on', 'stby', 'alt'] or numerical values 0-3. 
 - `sagetech_mxs squawk <squawk>` : The user may set the squawk code to an octal value between 0 and 7777.
 - `sagetech_mxs ident`: This command sends the ident bit for 8 seconds in ADSB-out messages (an alternative to using the parameter).

## Manual Configuration

Manual configuration can be done by enabling __MXS_EXT_CFG__ in step 2.

In this mode the MXS can be configured using Sagetech Mission Control software for a more custom configuration.

1. Follow steps 1 and 2 setting __MXS_EXT_CFG__ to enabled.
2. Disconnect the MXS from the flight controller and connect to the MXS with Mission Control.
3. Set desired settings through the installation configuration panel (i.e. ICAO, emitter category, aircraft size, etc.) in Mission Control.
    > __NOTE:__ Ensure that the baud rate for the relevant COM port matches the __SER_MXS_BAUD__ parameter.
4. Reconnect the MXS to the flight controller and open QGroundControl.
5. Reboot the flight controller.

If all steps were followed the settings set in Mission Control should appear in the Transponder parameter tree.

## Further Documentation
Further Sagetech Documentation and projects can be found on our GitHub. Some useful links can be found below.
- [MXS Host Interface Control Document](https://github.com/Sagetech-Avionics/sagetech-mxs-sdk/blob/main/doc/pdf/ICD02373_MXS_Host_ICD.pdf)

### CODE REVIEW

**Comment by ryanjAA** [comment]
This is so awesome!

File: .vscode/configurationCache.log [hunk 1]
```diff
@@ -0,0 +1 @@
...x4_fmu-v5_uavcanv1","px4_fmu-v5x","px4_fmu-v5x_default","px4_fmu-v5x_rtps","px4_fmu-v5x_test","px4_fmu-v6u","px4_fmu-v6u_bootloader","px4_fmu-v6u_default","px4_fmu-v6x","px4_fmu-v6x_bootloader","px4_fmu-v6x_default","px4_io-v2","px4_io-v2_default","px4_metadata","px4_raspberrypi","px4_raspberrypi_default","px4_sitl","px4_sitl_default","px4_sitl_default-clang","px4_sitl_nolockstep","px4_sitl_replay","px4_sitl_rtps","px4_sitl_test","px4fmu_firmware","px4io_update","python_coverage","qgc_firmware","quick_check","raspberrypi_pico","raspberrypi_pico_default","rostest","rostest_run","scan-build","scumaker_pilotpi","scumaker_pilotpi_arm64","scumaker_pilotpi_default","shellcheck_all","sizes","spracing_h7extreme","spracing_h7extreme_default","submodulesclean","submodulesupdate","tests","tests_avoidance","tests_coverage","tests_integration","tests_integration_coverage","tests_mission","tests_mission_coverage","tests_offboard","uorb_graphs","update_px4_msgs","update_px4_ros_com","update_ros2_bridge","updateconfig","uvify_core","uvify_core_default","validate_module_configs"],"launchTargets":[],"customConfigurationProvider":{"workspaceBrowse":{"browsePath":[],"compilerArgs":[]},"fileIndex":[]}}
```
Comment (dagar) [inline .vscode/configurationCache.log:1]: delete

File: .vscode/dryrun.log [hunk 2]
```diff
@@ -0,0 +1,19 @@
+make --dry-run --always-make --keep-going --print-directory
```
Comment (dagar) [inline .vscode/dryrun.log:1]: delete

File: src/drivers/transponder/sagetech_mxs/parameters.c [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+ *
+ * This parameter defines the squawk code. Value should be between 0000 and 7777.
+ *
+ * @reboot_required false
+ * @min 0
+ * @max 7777
+ * @group Transponder
+ *
+ */
+PARAM_DEFINE_INT32(ADSB_SQUAWK, 1200);
```
Comment (dagar) [inline src/drivers/transponder/sagetech_mxs/parameters.c:53]: I'm not sure, maybe this should be an MXS parameter unless we move it somewhere to be system wide accessible?

File: src/drivers/transponder/sagetech_mxs/parameters.c [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+ *
+ * Configure the size of the vehicle.
+ *
+ * @reboot_required true
+ * @min 0
+ * @max 15
+ * @group Transponder
+ *
+ * @value 0 SizeUnknown
+ * @value 1 Len15_Wid23
```
Comment (dagar) [inline src/drivers/transponder/sagetech_mxs/parameters.c:102]: Are people going to know what these mean or could we improve the naming?

**Comment by dagar** [comment]
Is the SDK officially online anywhere? Should we include it as a submodule rather than taking a copy? I want to make sure we're in a position to update it later if necessary.

**Comment by ryanjAA** [comment]
Sdk 

https://github.com/Sagetech-Avionics/sagetech-mxs-sdk

File: src/drivers/transponder/sagetech_mxs/parameters.c (hunk 3 above)
Comment (faberc) [reply src/drivers/transponder/sagetech_mxs/parameters.c:53]: This parameter and all the other parameters we have added in this driver will appear in the Transponders group. The squawk isn't MXS specific which is why we named it as ADSB_SQUAWK. This naming convention was primarily done to keep things consistent between the PX4 Driver and the Ardupilot Driver. Ardupilot implements several transponders, so the ADSB_ parameters apply to all of them.

File: .vscode/configurationCache.log (hunk 1 above)
Comment (faberc) [reply .vscode/configurationCache.log]: deleted

File: .vscode/dryrun.log (hunk 2 above)
Comment (faberc) [reply .vscode/dryrun.log]: deleted

File: .vscode/targets.log [hunk 5]
```diff
@@ -0,0 +1,2467 @@
+make all --print-data-base --no-builtin-variables --no-builtin-rules --question
```
[the remark this answers is not in the document]
Comment (faberc) [reply .vscode/targets.log]: deleted

File: src/drivers/transponder/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,34 @@
+############################################################################
+#
+#   Copyright (c) 2017 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/transponder/CMakeLists.txt:3]:
```suggestion
#   Copyright (c) 2022 PX4 Development Team. All rights reserved.
```

File: src/drivers/transponder/sagetech_mxs/parameters.c [hunk 7]
```diff
@@ -0,0 +1,269 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2017 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/transponder/sagetech_mxs/parameters.c:3]:
```suggestion
 *   Copyright (c) 2022 PX4 Development Team. All rights reserved.
```

**Comment by dagar** [comment]
I have no ability to test this, but given it doesn't run by default and isn't included in most boards this looks safe to merge to get more testing.

Once it's been validated let's do a pass to get it included everywhere, or at least all the boards with 2 MB of flash.

**Review by dagar — APPROVED** [review]

**Comment by ryanjAA** [comment]
> I have no ability to test this, but given it doesn't run by default and isn't included in most boards this looks safe to merge to get more testing.
> 
> Once it's been validated let's do a pass to get it included everywhere, or at least all the boards with 2 MB of flash.

I’ll have some of these soon so will test.

**Pull request merged**